src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStack.cs

```diff
@@ -208,6 +210,53 @@ public void Pop(bool success)
             SetConstructorArgumentState();
         }
 
+        /// <summary>
+        /// Configures the current stack frame for a polymorphic converter.
+        /// </summary>
+        public JsonConverter InitializePolymorphicReEntry(JsonTypeInfo derivedJsonTypeInfo)
+        {
+            Debug.Assert(!IsContinuation);
+            Debug.Assert(Current.PolymorphicJsonTypeInfo == null);
+            Debug.Assert(Current.PolymorphicSerializationState == PolymorphicSerializationState.None);
+
+            Current.PolymorphicJsonTypeInfo = Current.JsonTypeInfo;
+            Current.JsonTypeInfo = derivedJsonTypeInfo.PropertyInfoForTypeInfo.JsonTypeInfo;
+            Current.JsonPropertyInfo = Current.JsonTypeInfo.PropertyInfoForTypeInfo;
+            Current.NumberHandling ??= Current.JsonPropertyInfo.NumberHandling;
+            Current.PolymorphicSerializationState = PolymorphicSerializationState.PolymorphicReEntryStarted;
+            SetConstructorArgumentState();
+
+            return derivedJsonTypeInfo.PropertyInfoForTypeInfo.ConverterBase;
+        }
+
+
+        /// <summary>
+        /// Configures the current frame for a continuation of a polymorphic converter.
+        /// </summary>
+        public JsonConverter ResumePolymorphicReEntry()
+        {
+            Debug.Assert(Current.PolymorphicJsonTypeInfo != null);
+            Debug.Assert(Current.PolymorphicSerializationState == PolymorphicSerializationState.PolymorphicReEntrySuspended);
+
+            // Swap out the two values as we resume the polymorphic converter
+            (Current.JsonTypeInfo, Current.PolymorphicJsonTypeInfo) = (Current.PolymorphicJsonTypeInfo, Current.JsonTypeInfo);
```

Comment:
does that mean the `PolymorphicJsonTypeInfo` changes semantic meaning to `OriginalJsonTypeInfo` here? can you add a comment why we do this?

Reply:
Motivation is minimizing the amount of additional reference type fields we're adding to the stackframe struct. Doing this without swapping would require two additional pointers, which would contribute to the further bloating of the struct. The concern is that this would make the `Push()` and `Pop()` operations more expensive in a non pay-for-play way. I'll see if I can add a comment explaining this.